src/services/DisposableBase.ts

```diff
@@ -0,0 +1,55 @@
+import * as vscode from 'vscode';
+import type { IDisposable } from '../types';
+import { Logger } from '../util';
+
+/** Base class for disposing of dependencies. */
+export abstract class DisposableBase implements IDisposable {
+  constructor() {
+    this._logger = new Logger(`${this.constructor.name}: DisposableBase`);
+  }
+
+  protected readonly disposables = new Set<
+    (() => void) | IDisposable | vscode.Disposable
+  >();
+  private readonly _logger: Logger;
+
+  private _isDisposed = false;
+  private _isDisposing = false;
+  private _disposalPromise: Promise<any> | undefined;
+
+  /**
+   * Dispose of dependencies. Returns immediately if already called.
+   * @returns Promise that resolves when disposal is complete.
+   */
+  async dispose(): Promise<void> {
+    if (this._isDisposing) {
```

Comment:
Is there a point of differentiating `_isDisposing` and `_isDisposed`? Assuming you did this for a reason/ran into something.
Otherwise you could just check for `_disposalPromise` and return that.